overlord/devicestate/devicestate_serial_test.go

```diff
@@ -2408,3 +2409,194 @@ func (s *deviceMgrSerialSuite) TestDeviceSerialRestoreHappy(c *C) {
 	c.Check(log.String(), testutil.Contains,
 		fmt.Sprintf("restored serial serial-1234 for my-brand/pc-20 signed with key %v", devKey.PublicKey().ID()))
 }
+
+func (s *deviceMgrSerialSuite) TestShouldRequestSerial(c *C) {
+	type testCase struct {
+		deviceServiceAccess string
+		deviceServiceURL    string
+		storeAccess         string
+		gadgetName          string
+		expected            bool
+	}
+
+	testCases := []testCase{
+		{
+			storeAccess:         "",
+			gadgetName:          "",
+			deviceServiceAccess: "",
+			deviceServiceURL:    "",
+			expected:            true,
+		},
+		{
+			storeAccess:         "offline",
+			gadgetName:          "",
+			deviceServiceAccess: "",
+			deviceServiceURL:    "",
+			expected:            false,
+		},
+		{
+			storeAccess:         "",
+			gadgetName:          "gadget",
+			deviceServiceAccess: "offline",
+			deviceServiceURL:    "",
+			expected:            false,
+		},
+		{
+			storeAccess:         "",
+			gadgetName:          "gadget",
+			deviceServiceAccess: "",
+			deviceServiceURL:    "",
+			expected:            true,
+		},
+		{
+			storeAccess:         "offline",
+			gadgetName:          "gadget",
+			deviceServiceAccess: "",
+			deviceServiceURL:    "https://example.com",
+			expected:            true,
+		},
+		{
+			storeAccess:         "offline",
+			gadgetName:          "gadget",
+			deviceServiceAccess: "",
+			deviceServiceURL:    "",
+			expected:            false,
+		},
+		{
+			storeAccess:         "offline",
+			gadgetName:          "gadget",
+			deviceServiceAccess: "",
+			deviceServiceURL:    "https://example.com",
+			expected:            true,
+		},
+	}
+
+	s.state.Lock()
+	defer s.state.Unlock()
+
+	for _, t := range testCases {
+		tr := config.NewTransaction(s.state)
+		err := tr.Set("core", "store.access", t.storeAccess)
+		c.Assert(err, IsNil)
+
+		if t.gadgetName != "" {
+			err = tr.Set(t.gadgetName, "device-service.access", t.deviceServiceAccess)
+			c.Assert(err, IsNil)
+			err = tr.Set(t.gadgetName, "device-service.url", t.deviceServiceURL)
+			c.Assert(err, IsNil)
+		}
+		tr.Commit()
+
+		shouldReqest, err := devicestate.ShouldRequestSerial(s.state, t.gadgetName)
+		c.Check(err, IsNil)
+		c.Check(shouldReqest, Equals, t.expected, Commentf("%+v", t))
```

Comment:
If this fails, the output will be hard to parse. It would be easier to understand if the test cases had a name that we could use to print and match to a test. If the test names would be too awkward or hard to come up with, then just printing the index of the failed tests would be fine